File: hackerrank/problem solving/algorithms/1 - warmup/4 - diagonal difference/diagonal_difference.rb
# https://www.hackerrank.com/challenges/diagonal-difference/problem?isFullScreen=true

def main
  n = gets.to_i
  puts diagonal_difference(n)
end

# T: O(n^2)
# S: O(n) extra space
def diagonal_difference(n)
  primary_sum, secondary_sum = 0, 0

  n.times do |i|
    arr = read_numbers
    primary_sum += arr[i]
    secondary_sum += arr[-i - 1]
  end

  (primary_sum - secondary_sum).abs
end

def read_numbers
  gets.split.map!(&:to_i)
end

main if __FILE__ == $0

File: hackerrank/problem solving/algorithms/1 - warmup/4 - diagonal difference/diagonal_difference.py
# https://www.hackerrank.com/challenges/diagonal-difference/problem?isFullScreen=true

def main():
    n = int(input())
    print(diagonal_difference(n))


# T: O(n^2)
# S: O(n) extra space
def diagonal_difference(n):
    primary_sum, secondary_sum = 0, 0

    for i in range(n):
        arr = read_numbers()
        primary_sum += arr[i]
        secondary_sum += arr[-i - 1]

    return abs(primary_sum - secondary_sum)


def read_numbers():
    return list(map(int, input().split()))


if __name__ == '__main__':
    main()
